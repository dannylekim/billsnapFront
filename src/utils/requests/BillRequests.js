const { URL } = require("../../config");
const { BILLSNAP_TOKEN, BILLS } = require("../../constants/constants");
/**
 * @function getBill
 * @description Function that calls the /bills route to return the bills associated to the user. With bearer token.
 */
export const getBill = async () => {
  const token = localStorage.getItem(BILLSNAP_TOKEN);
  const response = await fetch(`${URL}/${BILLS}`, {
    method: "GET",
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });
  return response.json();
};

/**
 * @function createBill
 * @description function that creates a bill by calling the secured POST /bills endpoint.
 * @param createBillParams (which has name, category, company, items, accountsList, tipAmount, tipPercent, taxes)
 * @returns a Promise<Bill> on code 201
 */
export const createBill = async (createBillParams) => {
  const token = localStorage.getItem(BILLSNAP_TOKEN);
  const response = await fetch(`${URL}/${BILLS}`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json"
    },
    body: JSON.stringify(createBillParams),
  });
  return checkStatus(response);
};


async function checkStatus(response) {
  const parsedResponse = await response.json();
  if (parsedResponse.status !== "CREATED") {
    throw parsedResponse;
  }
  return parsedResponse;
}
