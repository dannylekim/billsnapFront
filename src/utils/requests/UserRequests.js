const { URL } = require("../../config")

/**
 * @function logUser
 * @description Function used to wrap the login request while simply requiring credentials as params
 * @param {String} credentials.email
 * @param {String} credentials.password // no need to encrypt yet
 */
export const login = async (credentials) => {

    const response = await fetch(`${URL}/login`, {
        method: "POST",
        headers: {
            "Content-Type": "application/json"
        },
        body: JSON.stringify({...credentials}),
    });

    return await checkStatus(response);
};

/**
 * @function registerUser
 * @description Function used to wrap the register request while simply requiring data for account registration
 * @param {String} entries.firstName
 * @param {String} entries.lastName
 * @param {String} entries.email
 * @param {String} entries.password // no need to encrypt yet
 */

/*
{type: "cors", url: "https://billsnap-development.herokuapp.com/billsnap/register", redirected: false, status: 400, ok: false, …}
*/
export const register = async (entries) => {

    const response = await fetch(`${URL}/register`, {
        method: "POST",
        headers: {
            "Content-Type": "application/json"
        },
        body: JSON.stringify({...entries}),
    });

    return await checkStatus(response);
};

async function checkStatus(response) {
    const parsedResponse = await response.json();
  
    if (response.status < 200 || response.status >= 300) {
        
        // throw new Error(parsedResponse.errors);
    }

    return {...parsedResponse, statusCode: response.status};
}
