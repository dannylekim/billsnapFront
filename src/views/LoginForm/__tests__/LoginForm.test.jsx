import React from "react";
import LoginFormContainer, {
  LoginForm,
  DEFAULT_ERRORS,
  DEFAULT_ALERT_MESSAGE,
} from "../LoginForm.jsx";
import { shallow, mount } from "enzyme";
import { login } from "../../../utils/requests/UserRequests";
jest.mock("../../../utils/requests/UserRequests");

describe("LoginForm", () => {
  let wrapper;
  let handleMockFunction,
    mockOnChange,
    mockDismissAlert,
    mockHandleResponse,
    mockHandleSubmitClick,
    mockToggleFormType,
    mockloginRequest;
  let setState, useStateSpy;

  beforeEach(() => {
    setState = jest.fn();
    useStateSpy = jest
      .spyOn(React, "useState")
      .mockImplementation((init) => [init, setState]);
    mockToggleFormType = jest.fn();
    handleMockFunction = jest.fn();
    mockDismissAlert = jest.fn();
    mockOnChange = jest.fn();
    mockHandleResponse = jest.fn();
    mockloginRequest = jest.fn();
    mockHandleSubmitClick = jest
      .fn()
      .mockImplementation(() => mockHandleResponse);


    wrapper = shallow(<LoginFormContainer setFormType={mockToggleFormType} />);
  });

  afterEach(() => {
    handleMockFunction.mockRestore();
    mockToggleFormType.mockRestore();
    mockDismissAlert.mockRestore();
    mockOnChange.mockRestore();
    mockHandleResponse.mockRestore();
    mockHandleSubmitClick.mockRestore();
    login.mockRestore();
  });

  describe("render", () => {
    describe("snapshots ðŸ“¸", () => {
      it("LoginFormWrapper should match snap shot", () => {
        const emailToolTip = document.createElement("div");
        const passwordToolTip = document.createElement("div");

        emailToolTip.setAttribute("id", "email");
        document.body.appendChild(emailToolTip);
        passwordToolTip.setAttribute("id", "password");
        document.body.appendChild(passwordToolTip);

        matches(
          <LoginForm
            onChange={mockOnChange}
            handleButtonClick={handleMockFunction}
            hasErrors={DEFAULT_ERRORS}
            alertMessage={DEFAULT_ALERT_MESSAGE}
            error_message=""
            dismissAlert={handleMockFunction}
          />
        );
      });

      it("LoginForm should match snap shot", () => {
        matches(<LoginFormContainer setFormType={mockToggleFormType} />);
      });
    });

    describe("components", () => {
      describe("ToolTip closed", () => {
        it("Should have no Errors ToolTip if closed", () => {
          wrapper = shallow(
            <LoginForm
              onChange={mockOnChange}
              setFormType={mockToggleFormType}
              handleButtonClick={mockHandleSubmitClick}
              hasErrors={DEFAULT_ERRORS}
              alertMessage={DEFAULT_ALERT_MESSAGE}
              error_message=""
              dismissAlert={mockDismissAlert}
            />
          );

          expect(wrapper.find("Alert")).toHaveLength(0);
          expect(wrapper.find("FormInput")).toHaveLength(2);
          expect(wrapper.find("FormInput").at(0).prop("invalid")).toBeFalsy();
          expect(wrapper.find("FormInput").at(1).prop("invalid")).toBeFalsy();

          expect(wrapper.find("Tooltip")).toHaveLength(2);
          expect(wrapper.find("Tooltip").at(0).prop("open")).toBeFalsy();
          expect(wrapper.find("Tooltip").at(1).prop("open")).toBeFalsy();
        });
      });

      describe("ToolTip opened", () => {
        it("Should have error, ToolTip is opened", () => {
          const error = {
            email: { hasError: true, message: "email error" },
            password: { hasError: true, message: "password error" },
          };

          const errorAlertMessage = {
            visible: true,
            alertType: "danger",
          };
          wrapper = shallow(
            <LoginForm
              onChange={mockOnChange}
              setFormType={mockToggleFormType}
              handleButtonClick={mockHandleSubmitClick}
              hasErrors={error}
              alertMessage={errorAlertMessage}
              error_message=""
              dismissAlert={mockDismissAlert}
            />
          );

          expect(wrapper.find("Alert")).toHaveLength(1);
          expect(wrapper.find("Alert").prop("open")).toBeTruthy();
          expect(wrapper.find("Alert").prop("theme")).toEqual("danger");

          expect(wrapper.find("FormInput")).toHaveLength(2);
          expect(wrapper.find("FormInput").at(0).prop("invalid")).toBeTruthy();
          expect(wrapper.find("FormInput").at(1).prop("invalid")).toBeTruthy();
          expect(wrapper.find("Tooltip")).toHaveLength(2);
          expect(wrapper.find("Tooltip").at(0).prop("open")).toBeTruthy();
          expect(wrapper.find("Tooltip").at(1).prop("open")).toBeTruthy();
        });
      });
    });
  });

  describe("functions", () => {
    it("dismissAlert should close alert message", () => {
      const errorAlertMessage = {
        visible: true,
        alertType: "danger",
      };

      wrapper.setState({ alertMessage: errorAlertMessage });
      wrapper.instance().dismissAlert();

      expect(wrapper.state().alertMessage).toEqual({
        alertType: "danger",
        visible: false,
      });
      expect(wrapper.state().hasErrors).toEqual(DEFAULT_ERRORS);
    });

    describe("handleResponse is called", () => {
      it("BAD_REQUEST response", () => {
        const BAD_REQUEST_RESPONSE = {
          status: "BAD_REQUEST",
          errors: [
            {
              field: "password",
              message: "does not matter ðŸ’©",
            },
          ],
          message: "BAD_REQUEST message",
        };

        wrapper.instance().handleErrorResponse(BAD_REQUEST_RESPONSE);
        expect(wrapper.state().alertMessage).toEqual({
          alertType: "danger",
          visible: true,
        });
        expect(wrapper.state().hasErrors).toEqual({
          ...DEFAULT_ERRORS,
          password: {
            hasError: true,
            message: BAD_REQUEST_RESPONSE.errors[0].message,
          },
        });
        expect(wrapper.state().error_message).toEqual(
          BAD_REQUEST_RESPONSE.message
        );
      });

      it("UNAUTHORIZED response", () => {
        const UNAUTHORIZED_RESPONSE = {
          status: "UNAUTHORIZED",
          message: "UNAUTHORIZED message",
        };

        wrapper.instance().handleErrorResponse(UNAUTHORIZED_RESPONSE);
        expect(wrapper.state().alertMessage).toEqual({
          alertType: "danger",
          visible: true,
        });
        expect(wrapper.state().hasErrors).toEqual({
          email: {
            hasError: true,
            message: "email might not exist in system.",
          },
          password: {
            hasError: true,
            message: "forgot your password?",
          },
        });
        expect(wrapper.state().error_message).toEqual(
          UNAUTHORIZED_RESPONSE.message
        );
      });
    });

    it("onFormChange is called when form input is changed", () => {
      const mockDOMevent = {
        target: {
          name: "email",
          value: "billSnapUser",
        },
      };
      wrapper.instance().onFormChange(mockDOMevent);
      expect(wrapper.state().user_credentials).toEqual({
        email: mockDOMevent.target.value,
        password: "",
      });
      expect(wrapper.state().hasErrors).toEqual(DEFAULT_ERRORS);
    });

    describe("handleSubmitClick", () => {
      it("should follow steps on passing request", async () => {
        const mockDOMevent = {
          preventDefault: jest.fn(),
        };
        const mockPushFunction = jest.fn();
        login.mockResolvedValue({ token: "billsnaptest " });

        wrapper.instance().dismissAlert = mockDismissAlert;
        wrapper.instance().setState = setState;
        // wrapper.instance().handleErrorResponse = mockHandleResponse;
        wrapper.setProps({
          history: {
            push: mockPushFunction,
          },
        });

        await wrapper.instance().handleSubmitClick(mockDOMevent);

        expect(mockDismissAlert).toBeCalledTimes(1);
        expect(login).toBeCalledTimes(1);
        expect(mockPushFunction).toBeCalledTimes(1);
        expect(setState).toBeCalledTimes(2);
      });

      it("should follow steps on passing request with no token", async () => {
        const mockDOMevent = {
          preventDefault: jest.fn(),
        };
        const mockPushFunction = jest.fn();
        login.mockResolvedValue({});

        wrapper.instance().handleErrorResponse = mockHandleResponse;
        wrapper.setProps({
          history: {
            push: mockPushFunction,
          },
        });

        await wrapper.instance().handleSubmitClick(mockDOMevent);

        expect(login).toBeCalledTimes(1);
        expect(mockPushFunction).toBeCalledTimes(0);
        expect(mockHandleResponse).toBeCalledTimes(1);
        
      });
      it("should follow steps on failing request", async () => {
        const mockDOMevent = {
          preventDefault: jest.fn(),
        };
        const ERROR_MESSAGE = 'haha it failed ðŸ’©';
        login.mockRejectedValue(ERROR_MESSAGE);

        wrapper.instance().handleErrorResponse = mockHandleResponse;

        await wrapper.instance().handleSubmitClick(mockDOMevent);

        expect(login).toBeCalledTimes(1);
        expect(mockHandleResponse).toHaveBeenCalledWith(ERROR_MESSAGE);
        
      });
    });

    it.skip("click formToggle", () => {
      wrapper.find(".form__toggle").simulate("click");
      expect(mockToggleFormType).toBeCalledTimes(1);
    });
  });
});
