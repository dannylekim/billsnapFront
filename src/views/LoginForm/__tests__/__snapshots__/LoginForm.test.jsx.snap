// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm render snapshots ðŸ“¸ Login should match snap shot 1`] = `
<div>
  <div
    className="login__form"
  >
    <div
      className="hidden__div"
    />
    <img
      alt="character logo"
      className="character__icon__image"
      src="./billSnapIcon.png"
    />
    <form
      className=""
    >
      <div
        className="form__inputs"
      >
        <div
          className="form-group"
          onChange={[MockFunction]}
        >
          <input
            autoComplete="username"
            className="register__login__inputs form-control"
            id="email"
            name="email"
            placeholder="@ Enter your email"
            type="email"
          />
        </div>
        <div
          className="form-group"
          onChange={[MockFunction]}
        >
          <input
            autoComplete="new-password"
            className="register__login__inputs form-control"
            id="password"
            name="password"
            placeholder="ðŸ”’ Enter a password"
            type="password"
          />
        </div>
      </div>
      <div
        className="form-group"
      >
        <a
          className="forgot__password"
          href="/#"
        >
           Forgot Password? 
        </a>
        <button
          className="login_register__submit__button btn btn-primary btn-sm btn-pill"
          name="submit"
          onClick={[Function]}
          type="button"
        >
          Log in
        </button>
      </div>
    </form>
  </div>
  <div>
    <div
      className="form__seperator"
    >
      <hr
        className="form__horizontal__line"
      />
      Or
      <hr
        className="form__horizontal__line"
      />
    </div>
    <div>
      <h6>
        New to Billsnap?
      </h6>
      <button
        className="form__toggle btn btn-primary"
        onClick={[Function]}
        type="button"
      >
        ðŸ‘‹ Create account.
      </button>
    </div>
  </div>
</div>
`;

exports[`LoginForm render snapshots ðŸ“¸ LoginContainer should match snap shot 1`] = `
<div
  className="login__container"
>
  <div>
    <div
      className="login__form"
    >
      <div
        className="hidden__div"
      />
      <img
        alt="character logo"
        className="character__icon__image"
        src="./billSnapIcon.png"
      />
      <form
        className=""
      >
        <div
          className="form__inputs"
        >
          <div
            className="form-group"
            onChange={[Function]}
          >
            <input
              autoComplete="username"
              className="register__login__inputs form-control"
              id="email"
              name="email"
              placeholder="@ Enter your email"
              type="email"
            />
          </div>
          <div
            className="form-group"
            onChange={[Function]}
          >
            <input
              autoComplete="new-password"
              className="register__login__inputs form-control"
              id="password"
              name="password"
              placeholder="ðŸ”’ Enter a password"
              type="password"
            />
          </div>
        </div>
        <div
          className="form-group"
        >
          <a
            className="forgot__password"
            href="/#"
          >
             Forgot Password? 
          </a>
          <button
            className="login_register__submit__button btn btn-primary btn-sm btn-pill"
            name="submit"
            onClick={[Function]}
            type="button"
          >
            Log in
          </button>
        </div>
      </form>
    </div>
    <div>
      <div
        className="form__seperator"
      >
        <hr
          className="form__horizontal__line"
        />
        Or
        <hr
          className="form__horizontal__line"
        />
      </div>
      <div>
        <h6>
          New to Billsnap?
        </h6>
        <button
          className="form__toggle btn btn-primary"
          onClick={[Function]}
          type="button"
        >
          ðŸ‘‹ Create account.
        </button>
      </div>
    </div>
  </div>
</div>
`;
